src/par_iter/collect/mod.rs

```diff
@@ -0,0 +1,38 @@
+use super::ExactParallelIterator;
+use std::isize;
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+mod consumer;
+use self::consumer::CollectConsumer;
+
+pub fn collect_into<PAR_ITER,T>(mut pi: PAR_ITER, v: &mut Vec<T>)
+    where PAR_ITER: ExactParallelIterator<Item=T>,
+          PAR_ITER: ExactParallelIterator,
+          T: Send,
+{
+    let len = pi.len();
+    assert!(len < (isize::MAX) as usize);
+
+    v.truncate(0); // clear any old data
+    v.reserve(len); // reserve enough space
+    let target = v.as_mut_ptr(); // get a raw ptr
+    let writes = AtomicUsize::new(0);
+    let consumer = unsafe { // assert that target..target+len is unique and writable
+        CollectConsumer::new(&writes, target, len)
+    };
+    pi.drive(consumer);
```

Comment:
Instead of this manual consumer, what if you use `from_raw_parts_mut` for an uninitialized slice, then
`slice.par_iter_mut().zip(pi).for_each(/* atomic inc, assign */)`

I'm not at home, so I can't try yet, but if this works it would simplify a lot...